### PR TITLE
Support batching messages in robust-watch

### PR DESCRIPTION
**Pull request by dstorrs**
First, the sleep in robust-watch gets moved to the bottom of the loop so that initial setup is faster.

More importantly, robust-watch gains a new optional argument:  #:batch? boolean?

The default is #f, meaning reporting behavior will be unchanged from before this PR.

If set to #t, all activity messages for that loop will be sent via a single async-channel-put instead of a -put for each message.  e.g., instead of reporting four separate changes, it will report this:

'(
(robust add <path:/foo/bar>)
(robust add <path:/foo/bar/baz.txt>)
(robust change <path:/foo/jaz.txt>)
(robust removed <path:/foo/quux.txt>)
)

### CODE REVIEW

**Comment by dstorrs** [comment]
Ping

File: robust-watch.rkt [hunk 1]
```diff
@@ -3,49 +3,52 @@
 ;; This module provides a cross-platform, polling based file watch.
 
 (require
-  racket/contract)
+ racket/contract
+ racket/list
+ racket/set
+ )
```
Comment (zyrolasting) [inline robust-watch.rkt:9]: Please put closing parens on the last line of an expression.

File: robust-watch.rkt [hunk 2]
```diff
@@ -3,49 +3,52 @@
...
 
 (provide
-  (contract-out
-    [robust-poll-milliseconds (parameter/c exact-positive-integer?)]
-    [robust-watch  (->* () (path-on-disk?) thread?)]))
+ (contract-out
+  [robust-poll-milliseconds (parameter/c exact-positive-integer?)]
+  [robust-watch  (->* () (path-on-disk? #:batch? boolean?) thread?)]))
```
Comment (zyrolasting) [inline robust-watch.rkt:14]: Consider `any/c` instead of `boolean?`. That way people can use `and`, `member`, etc. without needing to coerce the computed value.

File: robust-watch.rkt [hunk 3]
```diff
@@ -54,43 +57,61 @@
...
-      (if (path-on-disk? complete)
-        (let ([next (get-robust-state complete)])
-          (hash-for-each
-            (mark-status state next)
-            (lambda (affected op)
-              (unless (equal? op 'same)
-                (report-activity op affected))))
-          (loop next))
-        (report-activity 'remove complete))))))
+            (define initial (get-robust-state complete))
```
Comment (zyrolasting) [inline robust-watch.rkt:71]: I'm seeing a lot of `cond/else` instead of `if`. That, and the complexity has grown enough to make this a little hard to follow. Are you open to refactoring this for readability?

**Comment by zyrolasting** [comment]
Added food for thought: The intensive and apathetic models also have a problem re: competing messages. My only concern about this change is that it could apply to every model, but for some reason it only applies to the robust model.

That being said, I think this can be useful to more people. But I think a better approach might be to hook into the channel itself, regardless of the model, and have it accumulate events (keeping only the latest X events for a file). Reading from the channel would consume all events accumulated thus far.

Does that seem better? If so, I am happy to put that change in.

**Comment by dstorrs** [comment]
On Fri, May 29, 2020 at 5:23 PM Sage Gerard <notifications@github.com>
wrote:

> Added food for thought: The intensive and apathetic models also have a
> problem re: competing messages. My only concern about this change is that
> it could apply to every model, but for some reason it only applies to the
> robust model.
>
> That being said, I think this can be useful to more people. But I think a
> better approach might be to hook into the channel itself, regardless of the
> model, and have it accumulate events (keeping only the latest X events for
> a file). Reading from the channel would consume all events accumulated thus
> far.
>
> Does that seem better? If so, I am happy to put that change in.
>

Hi Sage,

That would certainly work.  It's not exactly the same -- the PR I submitted
means that all events *from a certain update* are sent at once, whereas
what you're talking about would provide all the events *since the last time
the channel was drained*.  There could also be issues where events would be
lost if the channel weren't drained soon enough, since you said that you
were only going to keep the latest X events.  That won't matter when
someone is only interested in the current state of the file, but it *would*
matter if they were interested in keeping an accurate history.

Regarding the intensive model, I'm concerned about its real-world
usability, for two reasons:

1) According to the docs it does not monitor directories, only files.
2) If there are, e.g. 1000 files in the directory then that's 1000
threads.  I haven't explored how Racket handles large numbers of threads,
but I would be concerned about the scheduler handling large numbers of
threads in a timely way.

Dave

PS:  Am I understanding correctly that, under the apathetic watch model, if
you start watching /tmp/foo/ then a change to
/tmp/foo/bar/baz/quux/testfile would be reported as a change to /tmp/foo/ ?

**Comment by dstorrs** [comment]
I implemented your requests about:

1) closing parens on the same line
2) any/c vs boolean?
3) avoiding cond/else instead of if (I refactored it into a single cond as I think that's clearer than nested ifs)

I'm using Emacs with racket-mode, which unfortunately means that my indentation is hardcoded and there's no easy way for me to change it.  I would suggest bringing the PR into a new branch, running `git diff --ignore-space-at-eol -b -w` in order to find the changes while ignoring whitespace, and reindenting it in your preferred style if you like the PR.

Unfortunately, I accidentally did the fixes in a new branch that has one semantic change (it checks the file size as well as permissions and access time).  I think the best solution is to close this PR and I'll submit a new one.  Is that okay?  If not then I can try to manually move the requisite changes onto my master branch and update this PR but it would be a bit of a hairball.  I'll submit the other PR and you can decide if you want it.

**Comment by zyrolasting** [comment]
My question was if this change could adapt to benefit every model. I moved the discussions re: the merits of the models to #8 and #9. I'm open to changing the apathetic model to report the exact file changed to make it more conducive to batched messages.

>  I'm using Emacs with racket-mode, which unfortunately means that my indentation is hardcoded and there's no easy way for me to change it.

If you head back to your fork's `master` branch (assuming `upstream` points to this repository), this session will update your staging area to only the changes that lack whitespace.

```
# Set staging area to use code before change.
git diff HEAD upstream/master | git apply -

git add .

# Unstage whitespace.
git diff --cached -w | git apply --cached -R
```

**Comment by dstorrs** [comment]
On Mon, Jun 1, 2020 at 2:34 PM Sage Gerard <notifications@github.com> wrote:

> My question was if this change could adapt to benefit every model. I moved
> the discussions re: the merits of the models to #8
> <https://github.com/zyrolasting/file-watchers/issues/8> and #9
> <https://github.com/zyrolasting/file-watchers/issues/9>. I'm open to
> changing the apathetic model to report the exact file changed to make it
> more conducive to batched messages.
>
It sounds like there's two changes being discussed:

1) Make batch-mode available for all models
2) Make apathetic mode report individual file changes

I'd be inclined to do that in separate PRs -- any objections?

I'm using Emacs with racket-mode, which unfortunately means that my
> indentation is hardcoded and there's no easy way for me to change it.
>
> If you head back to your fork's master branch (assuming upstream points
> to this repository), this session will update your staging area to only the
> changes that lack whitespace.
>
> # Set staging area to use code before change.
> git diff HEAD upstream/master | git apply -
>
> git add .
>
> # Unstage whitespace.
> git diff --cached -w | git apply --cached -R
>
>
Oh, sweet.  That is insanely useful, thank you.  I'm still at work but will
put together a fresh PR later.

**Comment by zyrolasting** [comment]
Since this code seems to be abandoned in favor of #7, I'm moving the discussion there and closing this one out.
